Sanitise environment before calling out to system shell

peter-b opened the pull request:
On OS X Yosemite, the system sets the environment variable `$PATH` multiple times.  This causes undefined behaviour in software that uses the `$PATH` environment variable, because the POSIX specification does not specify which value should be used.

In particular, it was found that `put ":/usr/local/bin" after $PATH; put shell("env")` showed that the LiveCode was modifying the first occurrence of $PATH in the environment, but `bash` was using the last.

To address this, use execle(3) rather than execl(3) and provide an explicit environment variable array using the environment values from LiveCode's global variables.

peter-b committed:
[[feature-sanitiseenv]] Linux: Use explicit environment in shell call.

When calling out to the shell, always generate and pass an explicit
environment instead of inheriting LiveCode's.

peter-b committed:
[[feature-sanitiseenv]] Add MCU_environmentarray().

Add a new utility function that constructs a NULL-terminated C array
of environment "name=value" strings in a specified encoding from all
of the current global variables that start with "$".

peter-b committed:
[[feature-sanitiseenv]] OS X: Use explicit environment in shell call.

When calling out to the shell, always generate and pass an explicit
environment instead of inheriting LiveCode's.

peter-b commented:
Arguably, we should consider doing this in every place we might run a subprocess, on all platforms... :wink:

runrevmark commented on engine/src/util.cpp:
```diff
@@ -2960,6 +2960,83 @@ bool MCString::split(char p_char, MCString& r_head, MCString& r_tail)
+		 * MCVariable::is_env is protected, this is the only way to
+		 * check if we're supposed to treat the variable as an
+		 * environment variable. */
+		if (!MCStringBeginsWithCString (MCNameGetString (t_var->getname()),
```
I'd suggest just adding an accessor to the MCVariable class here 'isenv()' or some such - rather than checking like this

runrevmark commented on engine/src/dskmac.cpp:
```diff
@@ -7205,7 +7205,12 @@ struct MCMacDesktop: public MCSystemInterface, public MCMacSystemService
+                    
+                    char **t_env = { NULL };
+                    uindex_t t_envc = 0;
+                    /* UNCHECKED */ MCU_environmentarray(kMCStringEncodingNative, t_env, t_envc);
```
I think this should be UTF-8 encoding.

peter-b commented on engine/src/util.cpp:
```diff
@@ -2960,6 +2960,83 @@ bool MCString::split(char p_char, MCString& r_head, MCString& r_tail)
+										kMCStringOptionCompareExact))
+			continue;
+
+		switch (MCValueGetTypeCode (t_var->getvalueref()))
```
I'd like to remove this switch statement -- it seems like it _should_ be unnecessary.

peter-b committed:
[[ feature-sanitiseenv ]] Add MCVariable::isenv() accessor.

peter-b committed:
[[ feature-sanitiseenv ]] OS X: Use UTF-8 encoding for env variables.

peter-b committed:
[[ feature-sanitiseenv ]] Use MCVariable::isenv() instead of test for "$".

peter-b committed:
[[ feature-sanitiseenv ]] Remove MCValueRef type check.

Rely on the assertion when we try to format the variable.

livecodepanos committed:
[[ LC 6.7 GM1 ]] Tweaks to get rid of compile errors for ios 5.0 simulator/device

runrevmark committed:
Merge branch 'refactor' into develop

runrevmark committed:
Updated submodule pointers after merging refactor -> develop (7.0.0).

peter-b committed:
[[ Bug 13622 ]] Revert patch to deduplicate env variables before exec'ing.

This reverts commit eb311763c79b to avoid conflicts.

It also reverts corresponding changes made in commit 487718cb7081
to add the functionality to the 7.0.0 branch.

peter-b committed:
Merge branch 'feature-sanitiseenv-merge' into feature-sanitiseenv

peter-b commented:
Closing this; I will resubmit as PR for the `develop` branch.
